### PR TITLE
census cell dup check

### PR DESCRIPTION
**Pull request by bkmartinjr**
Initial commit of a notebook/tool which searches for potentially mislabeled copies of duplicate cells, i.e., those with potentially incorrect `is_primary_data` metadata.

PR also includes a change to the pre-commit configuration to lint items in the tools directory beyond the builder.

### CODE REVIEW

**Comment by bkmartinjr** [comment]
FYI: @jahilton

**Review by atolopko-czi — APPROVED** [review]
Looks solid and very interesting! And a masterclass in Python generator usage.

It will be fun to make the reporting more substantial in the future. A few ideas: 
- add dataset titles
- show total cell counts for dataset-oriented tables
- a report for the most commonly repeated cells (desc by dup counts)

It would also be interesting to see how may dup cells are false positives due to having low nnz counts, and thus have information-poor "fingerprints" and just match by chance. I suppose we could report nnz counts for dup cells, if that seems fruitful.

**Review by jahilton — COMMENTED** [review]
Understanding which datasets are colliding will dramatically speed up the investigation into the cause (most are either all within a Collection or have matching cells counts, but the remainder need the additional info).
1 thought - groupby the `obs_duplicate_primary` df on hash & collapse the dataset_id into an ordered list. Then grab a value_counts of each dataset_id list.

Also, include the collection_id if readily available

File: tools/cell_dup_check/finddups.ipynb [hunk 1]
```diff
@@ -0,0 +1,558 @@
...
+    "            .to_pandas()\n",
+    "            .set_index(\"soma_joinid\")\n",
+    "        )\n",
+    "        hashes = pd.Series(data=np.full((len(obs_df),), \"\"), index=obs_df.index)\n",
+    "\n",
+    "        for (obs_soma_joinids_chunk, _), X_chunk in X_sparse_iter(\n",
+    "            query, X_name=\"raw\", row_stride=row_stride\n",
+    "        ):\n",
+    "            for r, row_soma_joinid in enumerate(obs_soma_joinids_chunk):\n",
+    "                row_counts = X_chunk.getrow(r)\n",
```
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply tools/cell_dup_check/finddups.ipynb:106]: how about `X_row`?

**Comment by bkmartinjr** [comment]
> include the collection_id if readily available

@jahilton - the Census has no knowledge of collections and I've been repeatedly waved off from including it.

@brianraymor - is there a REST API that I can hit to map a dataset_id to a collection_id?

File: tools/cell_dup_check/finddups.ipynb [hunk 2]
```diff
@@ -0,0 +1,3029 @@
...
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "\"\"\"\n",
+    "For each duplicated hash (cell), find the datasets that overlap/contain the duplicate.\n",
+    "\n",
+    "In ther words, these are all of the datasets which have intersecting duplicate/primary hashes. \n",
```
Comment (ebezzi) [inline tools/cell_dup_check/finddups.ipynb:2992]:
```suggestion
    "In other words, these are all of the datasets which have intersecting duplicate/primary hashes. \n",
```

**Review by ebezzi — APPROVED** [review]
LGTM. One typo correction left in comments.

**Pull request merged**